src/background.js

```diff
@@ -212,15 +212,33 @@ async function setupContainer () {
   }
 }
 
-function reopenTab ({url, tab, cookieStoreId}) {
-  browser.tabs.create({
+async function maybeReopenTab (url, tab, request) {
+  const macAssigned = await getMACAssignment(url);
+  if (macAssigned) {
+    // We don't reopen MAC assigned urls
+    return;
+  }
+  const cookieStoreId = shouldContainInto(url, tab);
```

Comment:
Nit-picky note since we're refactoring here ...

I don't like that `shouldContainInto` returns 2 different variable types - i.e., a string of the `cookieStoreId` into which we should contain the request, or `false` if it doesn't need to be contained. But I'm not sure of a good way to refactor it, so I don't want to block the rest of this good code refactoring on it. But I'm thinking maybe something like ...


```js
function needsContaining(url, tab) {
  if (!url.startsWith("http")) {
    // we only handle URLs starting with http(s)
    return false;
  }

  if (isFacebookURL(url) && tab.cookieStoreId !== facebookCookieStoreId) {
    // Facebook-URL outside of Facebook Container Tab
    return true;
  } else if (!isFacebookURL(url) && tab.cookieStoreId === facebookCookieStoreId) {
    // Non-Facebook-URL inside Facebook Container Tab
    return true;
  }
  return false;
}

function whichCookieStoreForUrl(url) {
  if (isFacebookURL(url)) {
    return facebookCookieStoreId;
  } else {
    return "firefox-default";
  }
}
```

Then change this calling code to:
```js
if (!needsContaining(url, tab)) {
  return;
}
const cookieStoreId = whichCookieStoreForUrl(url);
```
?

Reply:
Thanks for the review. The proposed change needs multiple calls to `isFacebookURL`, which is something that I would like to avoid. Maybe, since the returned type mismatch is your valid concern, returning an empty string would work? Or we could move the logic inside `shouldContainInto` completely to the `maybeReopenTab` function, since it isn't called from other locations anymore?

Reply:
That might be better, but I've caught the functional programming bug to create the smallest, most meaningfully-named functions possible.

To do that, we could add memo-ization to the `isFacebookURL` function to remove the additional regex processing from the additional calls.

Reply:
I still approve the rest of this PR though, if you don't want to do this `shouldContainInto` refactor as part of this PR.

Reply:
I'm totally fine with doing more refactoring as part of this PR.

Remembering checked urls would be a nice optimization, also speeding up repeated visits to the same domains a little. Though, that would probably need additional logic to clean up / trim them, so we don't end up with an evergrowing list. Would you say that's still in the scope of a refactoring, and if yes, how should we approach the clean up?

Reply:
This is starting to feel like some additional refactoring to do in a follow-up. I filed https://github.com/mozilla/contain-facebook/issues/247 to tackle this separately.